cmd/juju/machine/remove.go

```diff
@@ -111,17 +124,27 @@ func (a removeMachineAdapter) destroyMachines(f func(...string) error, machines
 	return results, nil
 }
 
-func (c *removeCommand) getRemoveMachineAPI() (RemoveMachineAPI, error) {
-	if c.api != nil {
-		return c.api, nil
+func (c *removeCommand) getAPIRoot() (api.Connection, error) {
+	if c.apiRoot != nil {
+		return c.apiRoot, nil
 	}
-	root, err := c.NewAPIRoot()
+	return c.NewAPIRoot()
+}
+
+func (c *removeCommand) getRemoveMachineAPI() (RemoveMachineAPI, error) {
+	root, err := c.getAPIRoot()
 	if err != nil {
 		return nil, err
 	}
-	if root.BestFacadeVersion("MachineManager") >= 3 {
+	if root.BestFacadeVersion("MachineManager") < 4 && c.KeepInstance {
+		return nil, errors.New("this version of Juju doesn't support --keep-instance")
```

Comment:
I want to fail early before making an api call. Also, there will be a generic "Not Implemented" return from the client facade, but that could be for other reasons as well. Here is where we have the knowledge of what's been asked for and what's supported. This pattern is used in several other CLI commands as well. See other comment below also.